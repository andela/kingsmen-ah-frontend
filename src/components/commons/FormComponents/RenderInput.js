/* eslint-disable jsx-a11y/label-has-for */
import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './index.scss';

class RenderInput extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool, PropTypes.number]),
    label: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    className: PropTypes.string.isRequired,
    error: PropTypes.string,
    handleChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: '',
    error: '',
  };

  constructor(props) {
    super(props);

    this.state = {
      changed: false
    };
  }

  onFocus = () => {
    this.setState({
      changed: true
    });
  }

  onFocusOut = () => {
    const { value } = this.props;
    if (value === '') {
      this.setState({
        changed: false
      });
    }
  }

  render() {
    const {
      id, type, name, className,
      value, error, label, handleChange
    } = this.props;
    const { changed } = this.state;

    return (
      <Fragment>
        <div className="my-4 relative mb-8">
          <label htmlFor={id} className={classNames({ 'has-value': changed })}>{label}</label>

          <input
            id={id}
            type={type}
            name={name}
            className={`${className} w-full py-2 border-gray-600 border-b-2 outline-none`}
            value={value}
            onChange={handleChange}
            onFocus={this.onFocus}
            onBlur={this.onFocusOut}
            error={error}
          />
          {error && <div className="text-red-600 text-xs text-left mt-1">{error}</div>}
        </div>
      </Fragment>
    );
  }
}

export default RenderInput;
