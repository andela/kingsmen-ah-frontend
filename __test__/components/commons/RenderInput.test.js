import React from 'react';
import RenderInput from '@components/commons/FormComponents/RenderInput';

const props = {
    name: 'name',
    label: 'Name',
    placeholder: 'Name',
    type: 'text',
    error: "",
    value: '',
    className: '',
    id: 'name',
    handleChange: jest.fn(),
    onFocus: jest.fn(),
};

const mockChange = jest.fn();

describe('Render Input', () => {
    it('should render the RenderInput component when used', () => {
        const shallowWrapper = shallow(<RenderInput {...props} />);

        expect(toJson(shallowWrapper)).toMatchSnapshot();
        expect(shallowWrapper.find('RenderInput')).toBeTruthy();
    });

    it('should call onChange prop', () => {
      const wrapper = mount(<RenderInput {...props} value="new value" handleChange={mockChange} />);

      wrapper.find("input").simulate("change");
      expect(mockChange).toHaveBeenCalled();
      wrapper.unmount();
    });

    it('should call onFocus prop', () => {
      const wrapper = mount(<RenderInput {...props} onFocus={mockChange} />);

      wrapper.find("input").simulate("focus");
      expect(mockChange).toHaveBeenCalled();
    });

    it('should show validation error', () => {
      const wrapper = mount(<RenderInput {...props} error="Required" handleChange={mockChange} />);

      expect(wrapper.find('.text-red-600').text()).toEqual('Required');
    });
});
