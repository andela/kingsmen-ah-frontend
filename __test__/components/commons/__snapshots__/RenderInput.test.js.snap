// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Input should render the RenderInput component when used 1`] = `
<Fragment>
  <div
    className="my-4 relative mb-8"
  >
    <label
      className=""
      htmlFor="name"
    >
      Name
    </label>
    <input
      className=" w-full py-2 border-gray-600 border-b-2 outline-none"
      error=""
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      type="text"
      value=""
    />
  </div>
</Fragment>
`;
